Mathlib/Analysis/Calculus/Gradient.lean

```diff
@@ -0,0 +1,472 @@
+
+Such defintions are based on the definitions of `fderiv`, and can be transformed to each other
+for functions `f` from `Hilbert Space E` to `ℝ `.
+
+## Main results
+
+This file contains the following parts of gradient.
+* the definition of gradient.
+* the theorems translating between `HasGradnAt` and `HasFDerivAt`
+* equivalent theorems translating the `LittleO` notation in `HasGradnAt` to
+  the language of `ε - δ` convergence, easier to be calculated.
+* theorems of the basic calculations of gradient, which can be found as well for `fderiv`.
+* theorems connecting `HasGradnAt`, `ContinuousAt` and `ε - δ` convergence.
+* theorems showing that for the differiential function,
+  the gradient (if exists) at local minima is zero.
+-/
+
+open InnerProductSpace
+noncomputable section
+
+variable {E : Type*} [NormedAddCommGroup E] [InnerProductSpace ℝ E] [CompleteSpace E]
+variable {f: E → ℝ} {f': E → (E →L[ℝ] ℝ)} {x y : E} {s : Set E}
+variable {x y gradient : E} {gradient' : E →L[ℝ] ℝ}
+local notation "⟪" x ", " y "⟫" => @inner ℝ _ _ x y
+
+/-
+### Definition of gradient
+-/
+/-- Translate `g` in `Hilbert Space E` to its dual ℝ-ContinuousLinearMap : `⟨ g , ⬝ ⟩`-/
+local notation "∇*" gradient  => (toDualMap ℝ _) gradient
+/-- Translate the ℝ-ContinuousLinearMap on `Hilbert Space E` to
+  its dual element on E by Frechet-Riesz Represent Theorem -/
+def grad : (E →L[ℝ] ℝ) → E
+  | gradient => ((toDual ℝ E).symm gradient)
+
+theorem grad_equiv :
+  ∀ x y :E, inner (grad (f' x)) y = f' x y := by
+    intro x y; apply toDual_symm_apply
+
+/-- The gradient of a function `f` from Hilbert Space E to ℝ at a point `x`
+    is defined to be the dual of `fderiv` of `f` at `x` -/
+def gradn (f : E → ℝ) (x : E) : E := grad (fderiv ℝ f x)
+
+/-- The gradient of a function `f` from ℝ to ℝ at a point `x`
+    is defined to be the dual of `fderiv` of `f` at `x` -/
+def grad_one (f : ℝ → ℝ) (x : ℝ) : ℝ := (toDual ℝ ℝ).symm (fderiv ℝ f x)
+
+/-- The gradient of a function `f` from Hilbert Space E to ℝ at a point `x` is `grad`
+    defined to be `f` Has `FDeriv` at `x` with the dual ContinuousLinearMap `⟨ gead , ⬝ ⟩`,
+    which means `f y - f x = ⟨ gradf , y - x ⟩ + o (y - x)` -/
+def HasGradnAt (f : E → ℝ) (grad : E) (x : E) := HasFDerivAt f (∇* grad) x
+
+/-- The gradient of a function `f` from ℝ to ℝ at a point `x` is `grad`
+    defined to be `f` Has `FDeriv` at `x` with the dual ContinuousLinearMap `⟨ gead , ⬝ ⟩`,
+    which means `f y - f x = ⟨ gradf , y - x ⟩ + o (y - x)` -/
+def HasGradoneAt (f : ℝ → ℝ) (grad : ℝ) (x : ℝ) := HasFDerivAt f (∇* grad) x
+
+/-- The dual of the dual of an element in `(EuclideanSpace ℝ n) →L[ℝ] ℝ` is itself -/
+theorem toDual_of_toDual_eq_self (gradient': E →L[ℝ] ℝ): gradient' = ∇* (grad gradient'):= by
+  apply ContinuousLinearMap.ext_iff.mpr
+  intro x
+  have : (∇* (grad gradient')) x = ⟪grad gradient', x⟫ := rfl
+  rw [this]
+  apply toDual_symm_apply.symm
+
+/-
+### Gradient and its `ε - δ` definition
```

Comment:
In mathlib we state these kinds of result using filters, not `ε - δ`.  In principle you do this here, but I think it's better just to delete them all, because (as you will see from my review below) every result you prove about `HasGradnAt` is an immediate consequence of the corresponding result about `HasFDerivAt`, without needing to unfold definitions further.